pkg/relayer/miner/miner.go

```diff
@@ -0,0 +1,408 @@
+	"github.com/pokt-network/poktroll/pkg/observable/filter"
+	"github.com/pokt-network/poktroll/pkg/observable/logging"
+	"github.com/pokt-network/poktroll/pkg/relayer"
+	"github.com/pokt-network/poktroll/pkg/relayer/protocol"
+	servicetypes "github.com/pokt-network/poktroll/x/service/types"
+)
+
+var (
+	_             relayer.Miner = (*miner)(nil)
+	defaultHasher               = sha256.New()
+	// TODO_THIS_COMMIT: where (on-chain) should this come from?
+	// TODO_TECHDEBT: setting this to 0 to effectively disable mining for now.
+	// I.e. all relays are added to the tree.
+	defaultRelayDifficulty = 0
+)
+
+// miner implements the relayer.Miner interface.
+type miner struct {
+	hasher          hash.Hash
+	relayDifficulty int
+
+	// Injected dependencies
+	sessionManager relayer.RelayerSessionsManager
+	supplierClient client.SupplierClient
+	blockClient    client.BlockClient
+}
+
+// minedRelay is a wrapper around a relay that has been serialized and hashed.
+type minedRelay struct {
+	servicetypes.Relay
+	bytes []byte
+	hash  []byte
+}
+
+// NewMiner creates a new miner from the given dependencies and options. It
+// returns an error if it has not been sufficiently configured or supplied.
+func NewMiner(
+	deps depinject.Config,
+	opts ...relayer.MinerOption,
+) (*miner, error) {
+	mnr := &miner{}
+
+	if err := depinject.Inject(
+		deps,
+		&mnr.sessionManager,
+		&mnr.supplierClient,
+		&mnr.blockClient,
+	); err != nil {
+		return nil, err
+	}
+
+	for _, opt := range opts {
+		opt(mnr)
+	}
+
+	if err := mnr.validateConfigAndSetDefaults(); err != nil {
+		return nil, err
+	}
+
+	return mnr, nil
+}
+
+// MineRelays kicks off relay mining by mapping the servedRelays ovservable through
+// a pipeline which hashes the relay, checks if it's above the mining difficulty,
+// adds it to the session tree, and then maps any errors to a  new observable.
+// It also starts the claim and proof pipelines which are subsequently driven by
+// mapping over RelayerSessionsManager's SessionsToClaim return observable.
+// It does not block as map operations run in their own goroutines.
+func (mnr *miner) MineRelays(
+	ctx context.Context,
+	servedRelays observable.Observable[servicetypes.Relay],
+) {
+	// sessiontypes.Relay ==> either.Either[minedRelay]
+	eitherMinedRelays := mnr.mineRelays(ctx, servedRelays)
+
+	// either.Either[minedRelay] ==> error
+	miningErrors := mnr.addReplayToSessionTree(ctx, eitherMinedRelays)
+	logging.LogErrors(ctx, miningErrors)
+
+	claimedSessions := mnr.createClaims(ctx)
+
+	mnr.submitProofs(ctx, claimedSessions)
+}
+
+// createClaims maps over the RelaySessionsManager's SessionsToClaim return
+// observable. For each claim, it calculates and waits for the earliest block
+// height at which it is safe to claim and does so. It then maps any errors to
+// a new observable which are subsequently logged. It returns an observable of
+// the successfully claimed sessions. It does not block as map operations run
```

Comment:
Thoughts on adding `TODO_DOCUMENT: Add a diagram of the major functions and observable flows in the miner pkg` above?